src/ordering/db/buy_order.py

```diff
@@ -0,0 +1,129 @@
+    __tablename__ = "buy_orders"
+
+    _db_id: int = sa.Column("id", sa.Integer, primary_key=True)
+    id: UUID = sa.Column("order_id", sa_utils.UUIDType, unique=True)
+    request_id: UUID = sa.Column(sa_utils.UUIDType, unique=True)
+
+    paid: Decimal = sa.Column(FiatAmountColumn)
+    bought: Decimal = sa.Column(BtcAmountColumn)
+
+    _currency: Currency = sa.Column(
+        "currency", sa.Enum(Currency, naive_enum=False)
+    )
+    _price: Decimal = sa.Column("price", FiatAmountColumn)
+    _rate_date: datetime = sa.Column("rate_date", sa.DateTime)
+
+    @hybrid_property
+    def exchange_rate(self) -> BTCRate:
+        return BTCRate(
+            price=self._price,
+            currency=self._currency,
+            on_date=self._rate_date,
+        )
+
+    @exchange_rate.setter
+    def exchange_rate(self, value: BTCRate) -> None:
+        self._price = value.price
+        self._currency = value.currency
+        self._rate_date = value.on_date
+
+    when_created: datetime = sa.Column(
+        "when_created", sa.DateTime, default=datetime.utcnow, index=True
+    )
+    _when_updated: datetime = sa.Column(
+        "when_updated", sa.DateTime, onupdate=datetime.utcnow,
+    )
+
+    def __repr__(self) -> Text:
+        return (
+            "<"
+            f"{self.__class__.__name__}"
+            f" db_id={self._db_id}"
+            f" order_id={self.id}"
+            f" request_id={self.request_id}"
+            f" amount={self.bought!s}BTC"
+            ">"
+        )
+
+
+@inject
+class ORMRepository(Repository):
+    _session: Session
+    _pending_events: List[Event]
+
+    def __init__(self, transaction: Transaction, bus: EventBus) -> None:
+        self._transaction = transaction
+        self._bus = bus
+
+    @contextmanager
+    def lock(self) -> ContextManager[Decimal]:
+        events = []
+        with self._transaction() as session:
+            self._session = session
+            self._pending_events = events
+
+            lock = f"LOCK TABLE {DBBuyOrder.__tablename__} IN EXCLUSIVE MODE;"
+            session.execute(lock)
+
+            yield self._calculate_balance()
+
+            del self._session
+            del self._pending_events
+
+        for event in events:
+            self._bus.emit(event)
+
+    def _calculate_balance(self) -> Decimal:
+        balance, = self._session.query(func.sum(DBBuyOrder.bought)).one()
+        return Decimal(balance or 0)
+
+    def create(
+            self,
+            request_id: UUID,
+            paid: Decimal,
+            bought: Decimal,
+            with_rate: BTCRate,
+    ) -> BuyOrder:
+        entry = DBBuyOrder()
+        entry.id = uuid4()
+        entry.request_id = request_id
+        entry.paid = paid
+        entry.bought = bought
+        entry.exchange_rate = with_rate
+        self._session.add(entry)
+        self._session.flush()
+        return entry
+
+    def emit(self, event: Event) -> None:
+        self._pending_events.append(event)
```

Comment:
Interesting, so in your design repository emits events. It makes sense if that's your choice

Reply:
This is another interesting topic, and also I had some talks about this earlier.
I think it's most practical if it's in the repository. The responsibility of the repository is to have a cohesive state of aggregate. So it's in a transaction. If you send your event out of your system you need to implement an outbox pattern. Not emitted events should be saved in the same transaction as aggregate. So We need to do it in the repository.
Another approach is to use Event sourcing, but this will also require a repository to handle events. So, in my opinion, it should go through the repository even if the transaction is on a whole HTTP request, extraction of component from service, and adding outbox pattern should be easier.